src/main/java/org/shanerx/tradeshop/utils/Utils.java

```diff
@@ -266,29 +268,51 @@ public String colorize(String msg) {
 	 * @return the sign.
 	 */
 	public Sign findShopSign(Block chest) {
+		//region Linked Sign Check - Load and Check Chest Linkage first, if it is a shop then return Linked Shop
 		ShopLocation potentialLocation = PLUGIN.getDataStorage().getChestLinkage(new ShopLocation(chest.getLocation()));
 		if (potentialLocation != null && ShopType.isShop(potentialLocation.getLocation().getBlock()))
 			return (Sign) potentialLocation.getLocation().getBlock().getState();
+		//endregion
 
-		ArrayList<BlockFace> faces = PLUGIN.getListManager().getDirections(),
-				flatFaces = new ArrayList<>(Arrays.asList(BlockFace.NORTH, BlockFace.EAST, BlockFace.SOUTH, BlockFace.WEST));
+		// Directions to check for Sign(Reversed later as the list is built to check from Sign->Chest and this method checks from Chest->Sign)
+		ArrayList<RelativeDirection> directions = PLUGIN.getListManager().getDirections();
+		// Directions to check for other half of a DoubleChest
+		ArrayList<BlockFace> flatFaces = new ArrayList<>(Arrays.asList(BlockFace.NORTH, BlockFace.SOUTH, BlockFace.EAST, BlockFace.WEST));
 
-		for (BlockFace face : faces) {
-			face = face.getOppositeFace(); // Check in the opposite direction that a sign would check
-			Block relative = chest.getRelative(face);
+		Block otherHalf = null;
+
+		for (RelativeDirection direction : directions) {
+			// Translate Relative Direction and get the Block
+			LocationOffset relativeOffset = direction.getTranslatedDirection(RelativeDirection.getDirection(chest.getBlockData()).getOppositeFace());
+			Block relative = chest.getRelative(relativeOffset.getXOffset(), relativeOffset.getYOffset(), relativeOffset.getZOffset());
+
+			// Check if it is a shop and return. Add the chest and shop to ChestLinkage, If it is a DoubleChest also add the other half
 			if (ShopType.isShop(relative)) {
+				PLUGIN.getDataStorage().addChestLinkage(new ShopLocation(chest.getLocation()), new ShopLocation(relative.getLocation()));
 				if (ShopChest.isDoubleChest(chest)) {
 					PLUGIN.getDataStorage().addChestLinkage(new ShopLocation(ShopChest.getOtherHalfOfDoubleChest(chest).getLocation()), new ShopLocation(relative.getLocation()));
 				}
 				return (Sign) relative.getState();
-			} else if (flatFaces.contains(face) && (chest.getType().equals(Material.CHEST) || chest.getType().equals(Material.TRAPPED_CHEST))) {
+				// 	If the relative block wasn't the ShopSign and we are checking a FlatFace, Then check if it is the other half of this Chest
+			} else if (flatFaces.stream().anyMatch((face) -> relativeOffset.getXOffset() == face.getModX() ||
+					relativeOffset.getYOffset() == face.getModY() ||
+					relativeOffset.getZOffset() == face.getModZ()) && (chest.getType().equals(Material.CHEST) || chest.getType().equals(Material.TRAPPED_CHEST))) {
```

Comment:
I'm confused by what's going on here. Why are these materials hard-coded? Other things are.also possible...

Reply:
This code is specifically for double chests which(as far as I know) can only be Chest and Trapped_Chest